<!-- # Child Rmarkdown Document for 10X genomics cellranger-based scRNA sequencing outputs -->
<!-- Parent document must have a variable "rna_in_dir" containing a subfolder "labeled_h5" -->
<!-- Note: should this be changed to just sample_h5? to allow for other types of h5 files? ie metadata-injected h5 pre-hashing? -->

```{r}
scrna_seq_module_version <- "1.0.1" # 20211006 extend for ARC metadata columns

```

The following metrics summarize the sequencing and alignment by 10x well prior to un-hashing and hash-based cell filtering.  

#### Contents 

##### [Pool Summary](#pool_seq_summary)
##### [Detailed Well Summary](#well_seq_summary)
##### [Plots of Well-Level Metrics](#seq_metrics_plots)  


```{r scrna_seq_dependency_check, include = FALSE}
assertthat::assert_that(exists("rna_in_dir"))

# check that we have h5 files provided. Can accept any type where 10x metadata has been injected
h5_file_dir_options <- file.path(rna_in_dir, c("meta_h5","merged_h5","labeled_h5"))
assertthat::assert_that(any(dir.exists(h5_file_dir_options)))

# which set of h5 of file will be used. current priority is cellranger, then merged, then labeled
h5_file_dir <- h5_file_dir_options[dir.exists(h5_file_dir_options)][1]
assertthat::assert_that(length(dir(h5_file_dir)) > 0)

h5_file_type <- ifelse(h5_file_dir == h5_file_dir_options[1], "cellranger",
                       ifelse(h5_file_dir == h5_file_dir_options[2],"merged" ,
                              ifelse( h5_file_dir == h5_file_dir_options[3],"labeled", 
                                      stop("scrna seq module requires an h5 input folder in input directory"))))

```  

```{r scrna_seq_input_files}
# locate files unless already in memory from an upstream process. 
expected_files <- switch(h5_file_type,
                         cellranger = "all_cellranger_h5",
                         merged = "all_merged_h5",
                         labeled = "all_labeled_h5")

if(exists(expected_files)){
  all_h5 <- get(expected_files)
} else {
  all_h5 <- list.files(path = h5_file_dir, 
                               pattern = ".h5$", 
                               full.names = TRUE, recursive = TRUE)
}

if(!exists("well_info")){
  # Merge Well Data
  stm("Reading in labeled h5 file well data")
  
  well_list <- future_lapply(all_h5, read_h5_well_meta)
  well_info <- unique(do.call(rbind, well_list))
  setDT(well_info)
  well_info[, pool_id := gsub("C\\d+W\\d+", "", well_id)]
  
  remove("well_list")
}

```  
  
```{r scrna_definitions} 
vnames_rna <- c("estimated_number_of_cells", "fraction_reads_in_cells", 
            "mean_reads_per_cell", "median_genes_per_cell", "median_umi_counts_per_cell", 
            "number_of_reads", "q30_bases_in_barcode", "q30_bases_in_rna_read", 
            "q30_bases_in_sample_index", "q30_bases_in_umi", "reads_mapped_antisense_to_gene", 
            "reads_mapped_confidently_to_exonic_regions", "reads_mapped_confidently_to_genome", 
            "reads_mapped_confidently_to_intergenic_regions","reads_mapped_confidently_to_intronic_regions", 
            "reads_mapped_confidently_to_transcriptome", "reads_mapped_to_genome", 
            "sequencing_saturation", "total_genes_detected", "valid_barcodes")

vnames_arc <- c("estimated_number_of_cells",
  "gex_fraction_of_transcriptomic_reads_in_cells","gex_mean_raw_reads_per_cell","gex_median_genes_per_cell",
  "gex_median_umi_counts_per_cell","gex_percent_duplicates","gex_q30_bases_in_barcode",
  "gex_q30_bases_in_read_2","gex_q30_bases_in_sample_index_i1","gex_q30_bases_in_sample_index_i2",
  "gex_q30_bases_in_umi","gex_reads_mapped_antisense_to_gene","gex_reads_mapped_confidently_to_exonic_regions",
  "gex_reads_mapped_confidently_to_genome","gex_reads_mapped_confidently_to_intergenic_regions", "gex_reads_mapped_confidently_to_intronic_regions",
  "gex_reads_mapped_confidently_to_transcriptome","gex_reads_mapped_to_genome","gex_reads_with_tso",
  "gex_sequenced_read_pairs","gex_total_genes_detected","gex_valid_barcodes", "gex_valid_umis")  

if(all(vnames_rna %in% colnames(well_info))){
  vnames <- vnames_rna
  vlabels <- c("Estimated Number of Cells", "Fraction Reads in Cells", 
            "Mean Reads per Cell", "Median Genes per Cell", "Median UMI per Cell", 
            "Number of Reads", "Q30 Bases in Barcode (%)", "Q30 Bases in RNA Read (%)", 
            "Q30 Bases in Sample Index (%)", "Q30 Bases in UMI (%)", "Reads Mapped Antisense to Gene (%)", 
            "Reads Mapped Confidently to Exonic Regions (%)", "Reads Mapped Confidently to Genome (%)", 
            "Reads Mapped Confidently to Intergenic Regions (%)",
            "Reads Mapped Confidently to Intronic Regions (%)", 
            "Reads Mapped Confidently to Transcriptome (%)", "Reads Mapped to Genome (%)", 
            "Sequencing Saturation (%)", "Total Genes Detected", "Valid Barcodes (%)")
  n_vars <- length(vnames)
  vartypes <- c(rep("Cells", 5), rep("Sequencing", 5), rep("Mapping", 7),"Sequencing","Cells","Sequencing")
  vartypes <- factor(vartypes,  levels = c("Cells", "Sequencing", "Mapping"))

  digitsRound <- c(0, 1, rep(0, 4), rep(1, 12), 0, 1)
  rna_data_type <- "rna"
} else if (all(vnames_arc %in% colnames(well_info))){
  vnames <- vnames_arc
  vlabels <- c("Estimated Number of Cells","Fraction Transcriptomic Reads in Cells",
            "Mean Raw Reads per Cell", "Median Genes per Cell", "Median UMI per Cell", "Percent Duplicates",
            "Q30 Bases in Barcode", "Q30 Bases in RNA Read 2", "Q30 Bases in Sample Index i1",
            "Q30 Bases in Sample Index i2", "Q30 Bases in UMI", "Reads Mapped Antisense to Gene", 
            "Reads Mapped Confidently to Exonic Regions", "Reads Mapped Confidently to Genome", 
            "Reads Mapped Confidently to Intergenic Regions", 
            "Reads Mapped Confidently to Intronic Regions", 
            "Reads Mapped Confidently to Transcriptome", "Reads Mapped to Genome", "Reads with TSO",
            "Sequenced Read Pairs", "Total Genes Detected", "Valid Barcodes", "Valid UMIs") 
  n_vars <- length(vnames)
  vartypes <- c(rep("Cells", 5), rep("Sequencing", 6), rep("Mapping", 7),"Sequencing","Sequencing","Cells","Sequencing","Sequencing")
  digitsRound <- c(0, 3, 0, 0, 0, rep(3, 14), 0, 0, 3, 3)
  
  rna_data_type <- "arc"
} else {
    vnames <- setdiff( colnames(well_info), c("well_id","pool_id"))
    n_vars <- length(vnames)
    vlabels <- gsub("_"," ", vnames)
    vartypes <- rep(NA, n_vars)
    
    getDigits <- function(vNumeric){
      vChar <- as.character(vNumeric)
      is_decimal <- any(grepl("[.]", vChar))
      if(is_decimal){
        decimal_digits <- nchar(vChar) - (nchar(gsub("[.].*","", vChar))+1)
        n_digits <- max(decimal_digits)
      } else {
        n_digits <- 0
      }
      n_digits
    }
    digitsRound <- sapply(subset(well_info, select = vnames), getDigits)
    
    rna_data_type <- "unknown"
}

  df_vars <- data.frame(Category = vartypes,
                        Variable_name = vlabels,
                        Variable = vnames,
                        Round = digitsRound)
```


<a  id="pool_seq_summary"></a> 


#### Pool Summary    
```{r well_summary_table1, results = "asis"} 
stm("Creating scrna well cellranger summary table")

if(rna_data_type == "rna" | "number_of_reads" %in% names(well_info)){
  pool_info <- well_info %>%
  dplyr::group_by(pool_id) %>%
  dplyr::summarize(total_cells = formatC(sum(estimated_number_of_cells), big.mark = ",", digits = 0, format = "f"),
            total_reads = formatC(sum(number_of_reads), big.mark = ",", digits = 0, format = "f"), .groups = "drop")
} else if (rna_data_type == "arc" | "gex_sequenced_read_pairs" %in% names(well_info)){
  pool_info <- well_info %>%
  dplyr::group_by(pool_id) %>%
  dplyr::summarize(total_cells = formatC(sum(estimated_number_of_cells), big.mark = ",", digits = 0, format = "f"),
            total_sequenced_read_pairs = formatC(sum(gex_sequenced_read_pairs), big.mark = ",", digits = 0, format = "f"), .groups = "drop")
} else {
  print("Warning: No total reads column detected in well metadata")
  pool_info <- well_info %>%  
    dplyr::group_by(pool_id) %>%  
    dplyr::summarize(total_cells = formatC(sum(estimated_number_of_cells), big.mark = ",", digits = 0, format = "f"), 
                     total_reads = NA, .groups = "drop")
}

names(pool_info) <- stringr::str_to_title(gsub("_", " ", names(pool_info)))
pool_info %>%
  gt::gt() %>%
  gt::cols_align(align = "right", columns = 2:3)

```  
[Return to Top](#rna_seq_top) 

<a  id="well_seq_summary"></a>  
  
#### Detailed Well Summary  
```{r well_summary_table, results = "asis"}
unique_pools <- sort(unique(well_info$pool_id))

well_summary_table <- well_info %>% 
  gather(key = Variable, value = Value, all_of(vnames)) %>%   # all variables long
  full_join(df_vars, by = "Variable") %>%  
  group_by(pool_id, Category, Variable, Variable_name) %>% 
  summarize(Median = formatC(median(Value, na.rm=T), big.mark = ",", digits = unique(Round), format = "f"),
            Range = get_range(Value, digits = unique(Round), verbose = F), 
            `CV%` = round(sd(Value)/mean(Value)*100,1),
            .groups = "drop") %>% 
  arrange(Category, Variable_name) %>% 
  tidyr::pivot_wider(id_cols = c("Category","Variable","Variable_name"), 
                     names_from = pool_id, 
                     values_from = c("Median","Range","CV%"),
                     names_glue = "{pool_id}__{.value}",
                     names_sort = TRUE) %>%
  mutate(Plot = sprintf("[Plot](#%s)", Variable)) %>% 
  select(Category, Variable_name, contains(unique_pools), Plot) # reorder cols by pools first then stats, keep only the clean var name

# well_summary_table_comments <- flag_well_summary(well_summary_table)  # Eventual function to perform well-specific flagging based on summary values
well_summary_table_comments <- well_summary_table
# set.seed(1105)
# dummycomments <- sample(c("","Warning: message","Fail: message"),
#                         size = nrow(well_summary_table_comments), 
#                         replace=T, prob = c(0.85,0.12,0.03))

gt_table <- well_summary_table_comments %>% 
  gt::gt() %>% 
  gt::fmt_markdown(columns = "Plot") %>% # convert the plot column text to markdown to activate links. these links generated with plots below.
  gt::cols_width(vars(Category) ~ px(100),
                 vars(Variable_name) ~ px(150),
                 ends_with("Median") ~ px(100),
                 ends_with("Range") ~ px(100),
                 ends_with("CV%") ~ px(50),
                 vars(Plot) ~ px(60)) %>% 
  gt::tab_options(table.font.size = 11, column_labels.font.size = 12) %>%
  gt::tab_spanner_delim(delim = "__") %>% 
  gt::cols_align(align = "right",
                 columns = c(ends_with("Median"), ends_with("Range"))) %>% 
  gt::cols_align(align = "center",
                 columns = vars(Plot)) %>% 
  gt::cols_label(Variable_name = "Variable")

gt_table
  
```  

<details>
  <summary>Expand table of statistics per well</summary> 
```{r well_info_table}
qc_table(well_info)

```  

</details>   

[Return to Top](#rna_seq_top) 

<a  id="seq_metrics_plots"></a> 
  
#### Plots of Well-Level Metrics  
  
```{r well_plots, results = "asis", fig.height = 3.5}
stm("Generating sequencing and alignment QC plots")

verpal <- hcl.colors(n = n_vars, palette = "viridis")

# Plots
for (i in seq_along(vnames)){
  df <- data.table::copy(well_info)
  spec <- vnames[i]
  slabel <- vlabels[i]
  df <- as.data.frame(df)
  df$spec_col <- df[,spec]
  med_val <- median(df$spec_col)
  cv <- round(sd(df$spec_col)/mean(df$spec_col)*100, 2)
  n <- sum(!is.na(df$spec_col))
  
  g <- ggplot(df, aes(well_id, spec_col)) +
    geom_bar(stat = "identity", fill = verpal[i]) + 
    geom_hline(yintercept = med_val, linetype = "dashed", color = "red")+
    scale_y_continuous(sec.axis = dup_axis(breaks = med_val, labels = med_val, name = ""))+
    xlab("Well") +
    ylab(slabel) +
    facet_wrap(~pool_id, ncol = n_pools, scales = "free_x", drop = TRUE) +
    ggtitle(slabel, 
            subtitle = sprintf("Median=%s    CV=%.1f%%    N=%s", med_val, cv, n)) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
  
  # Plot-specific hyperlink definition
  cat(sprintf('\n<a id="%s"></a>', spec), labels = "", sep = "\n")
  
  # Output plot
  suppressWarnings(print(g))
  
  # Link back to top of section
  cat("  \n[Return to Top](#rna_seq_top)", labels = "", sep = "\n")
  
}

``` 
  
scRNA seq report module v.`r {scrna_seq_module_version}`
